File: source/F#/src/tarjan.fs
// Tarjan union-find for cycle counting

module Tarjan

type T =
   { seed : int array
     root : int array
     mutable sets : int }

let create n =
   { seed = Array.init n (fun x -> x)
     root = Array.zeroCreate n
     sets = n }

let reset t n =
   Array.blit t.seed 0 t.root 0 n
   t.sets <- n

let find t a =
   let mutable here = a

   while t.root.[here] <> here do
      here <- t.root.[here]

   let top = here
   let mutable here = a

   while t.root.[here] <> top do
      let next = t.root.[here]
      t.root.[here] <- top
      here <- next

   top

let unite t a b =
   let a = find t a
   let b = find t b

   if a <> b then
      t.sets <- t.sets - 1
      t.root.[a] <- b

let setCount t = t.sets

let name = "Tarjan"

File: source/haskell/src/Tarjan.hs
-- Tarjan union-find algorithm for cycle counting
{-# LANGUAGE Strict #-}

module Tarjan where

import Begin
import qualified Data.Vector.Unboxed.Mutable as MV

data Relations s = Relations
   { root :: STVector s Int
   , sets :: STRef s Int
   }

create :: Int -> ST s (Relations s)
create n = do
   root <- MV.generate n id
   sets <- newSTRef n
   return Relations{root, sets}

reset :: Relations s -> Int -> ST s ()
reset Relations{root, sets} n = do
   forM_ [0 .. n - 1] $ \i -> MV.write root i i
   writeSTRef sets n

find :: STVector s Int -> Int -> ST s Int
find root = go
  where
   go here = do
      parent <- MV.read root here
      if parent == here
         then return here
         else do
            top <- go parent
            MV.write root here top
            return top

unite :: Relations s -> Int -> Int -> ST s ()
unite Relations{root, sets} a b = do
   ra <- find root a
   rb <- find root b

   when (ra /= rb) $ do
      sets -= 1
      MV.write root ra rb

count :: Relations s -> ST s Int
count Relations{sets} = readSTRef sets
